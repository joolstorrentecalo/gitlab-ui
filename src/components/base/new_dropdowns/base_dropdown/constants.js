export const FIXED_WIDTH_CLASS_SMALL = 'gl-w-31!';
export const FIXED_WIDTH_CLASS_LARGE = 'gl-w-48!';
export const DEFAULT_OFFSET = 4;
