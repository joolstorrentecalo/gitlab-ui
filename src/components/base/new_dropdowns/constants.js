// base dropdown events
export const GL_DROPDOWN_SHOWN = 'shown';
export const GL_DROPDOWN_HIDDEN = 'hidden';
export const GL_DROPDOWN_FOCUS_CONTENT = 'focusContent';

// KEY Codes
export const ARROW_DOWN = 'ArrowDown';
export const ARROW_UP = 'ArrowUp';
export const END = 'End';
export const ENTER = 'Enter';
export const HOME = 'Home';
export const SPACE = 'Space';
