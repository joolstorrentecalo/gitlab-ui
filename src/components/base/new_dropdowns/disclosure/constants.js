export const DISCLOSURE_DROPDOWN_ITEM_NAME = 'GlDisclosureDropdownItem';
export const DISCLOSURE_DROPDOWN_GROUP_NAME = 'GlDisclosureDropdownGroup';
