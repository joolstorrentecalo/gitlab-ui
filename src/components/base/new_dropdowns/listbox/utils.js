import { isString, isNumber } from 'lodash';

const isOption = (item) => Boolean(item) && (isString(item.value) || isNumber(item.value));

const isGroup = ({ options } = {}) => Array.isArray(options) && options.every(isOption);

const hasNoDuplicates = (array) => array.length === new Set(array).size;

const flattenedOptions = (items) => items.flatMap((item) => (isOption(item) ? item : item.options));

const isAllOptionsOrAllGroups = (items) => items.every(isOption) || items.every(isGroup);

const hasUniqueValues = (items) =>
  hasNoDuplicates(flattenedOptions(items).map(({ value }) => value));

const hasUniqueGroups = (items) => hasNoDuplicates(items.filter(isGroup).map(({ text }) => text));

const itemsValidator = (items) =>
  isAllOptionsOrAllGroups(items) && hasUniqueValues(items) && hasUniqueGroups(items);

export { isOption, itemsValidator, flattenedOptions };
