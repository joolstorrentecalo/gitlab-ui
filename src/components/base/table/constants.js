export const tableFullSlots = [
  'bottom-row',
  'empty',
  'emptyfiltered',
  'table-busy',
  'thead-top',
  'top-row',
];

export const tableFullProps = [
  'api-url',
  'busy',
  'current-page',
  'empty-filtered-html',
  'empty-filtered-text',
  'empty-html',
  'empty-text',
  'filter',
  'filter-debounce',
  'filter-function',
  'filter-ignored-fields',
  'filter-included-fields',
  'label-sort-asc',
  'label-sort-clear',
  'label-sort-desc',
  'no-footer-sorting',
  'no-local-sorting',
  'no-provider-filtering',
  'no-provider-paging',
  'no-provider-sorting',
  'no-select-on-click',
  'no-sort-reset',
  'per-page',
  'select-mode',
  'selectable',
  'selected-variant',
  'show-empty',
  'sort-by',
  'sort-compare',
  'sort-compare-locale',
  'sort-compare-options',
  'sort-desc',
  'sort-direction',
  'sort-icon-left',
  'sort-null-last',
];
