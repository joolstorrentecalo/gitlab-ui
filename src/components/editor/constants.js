export const headingLevels = [1, 2, 3];
export const EDITOR_LITE_INSTANCE_ERROR_NO_EL = '"el" parameter is required for createInstance()';
export const URI_PREFIX = 'gitlab';
