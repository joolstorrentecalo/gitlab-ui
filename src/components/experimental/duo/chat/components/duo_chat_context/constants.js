export const CONTEXT_ITEM_TYPE_ISSUE = 'issue';
export const CONTEXT_ITEM_TYPE_MERGE_REQUEST = 'merge_request';
export const CONTEXT_ITEM_TYPE_PROJECT_FILE = 'project_file';
