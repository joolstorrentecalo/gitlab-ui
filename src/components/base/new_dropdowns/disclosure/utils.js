import { isFunction } from 'lodash';
import { DISCLOSURE_DROPDOWN_ITEM_NAME, DISCLOSURE_DROPDOWN_GROUP_NAME } from './constants';

const itemValidator = (item) => item?.text?.length > 0 && !Array.isArray(item?.items);

const isItem = (item) => Boolean(item) && itemValidator(item);

const isGroup = (group) =>
  Boolean(group) &&
  Array.isArray(group.items) &&
  Boolean(group.items.length) &&
  group.items.every(isItem);

const itemsValidator = (items) => items.every(isItem) || items.every(isGroup);

const isListItem = (tag) =>
  ['gl-disclosure-dropdown-group', 'gl-disclosure-dropdown-item', 'li'].includes(tag);

const isValidSlotTagVue2 = (vNode) =>
  Boolean(vNode) && isListItem(vNode.componentOptions?.tag || vNode.tag);

const isValidSlotTag = (vNode) => {
  return (
    [DISCLOSURE_DROPDOWN_ITEM_NAME, DISCLOSURE_DROPDOWN_GROUP_NAME].includes(vNode.type?.name) ||
    vNode.type === 'li'
  );
};

const hasOnlyListItems = ({ default: defaultSlot }) => {
  if (!isFunction(defaultSlot)) {
    return false;
  }
  const nodes = defaultSlot();
  return (
    Array.isArray(nodes) &&
    nodes.filter((vNode) => vNode.tag).length &&
    (nodes.filter((vNode) => vNode.tag).every(isValidSlotTagVue2) ||
      nodes.filter((vNode) => vNode.tag).every(isValidSlotTag))
  );
};

export { itemsValidator, isItem, isGroup, hasOnlyListItems };
