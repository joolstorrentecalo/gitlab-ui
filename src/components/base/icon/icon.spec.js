import { shallowMount } from '@vue/test-utils';
import { iconSizeOptions } from '~/utils/constants';
import Icon from './icon.vue';

const ICONS_PATH = '/path/to/icons.svg';
const TEST_SIZE = 8;
const TEST_NAME = 'check-circle';

jest.mock('@gitlab/svgs/dist/icons.svg', () => '/path/to/icons.svg');

describe('Icon component', () => {
  let wrapper;
  let consoleSpy;

  const createComponent = (props) => {
    wrapper = shallowMount(Icon, {
      propsData: {
        size: TEST_SIZE,
        name: TEST_NAME,
        ...props,
      },
    });
  };

  const validateSize = (size) => Icon.props.size.validator(size);
  const validateName = (name) => Icon.props.name.validator(name);

  afterEach(() => {
    if (consoleSpy) {
      consoleSpy.mockRestore();
    }
  });

  it('has role=img', () => {
    createComponent();

    expect(wrapper.attributes('role')).toBe('img');
  });

  describe('when created', () => {
    beforeEach(() => {
      createComponent();
    });

    it(`shows svg class "s${TEST_SIZE}" and path "${ICONS_PATH}#${TEST_NAME}"`, () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('size validator', () => {
    const maxSize = Math.max(...iconSizeOptions);

    it('fails with size outside options', () => {
      expect(validateSize(maxSize + 10)).toBe(false);
    });

    it('passes with size in options', () => {
      expect(validateSize(maxSize)).toBe(true);
    });
  });

  describe('name validator', () => {
    it('fails with name that does not exist', () => {
      const badName = `${TEST_NAME}-bogus-zebra`;
      consoleSpy = jest.spyOn(console, 'warn').mockImplementation();

      expect(validateName(badName)).toBe(false);

      expect(consoleSpy).toHaveBeenCalledWith(
        `Icon '${badName}' is not a known icon of @gitlab/svgs`
      );
    });

    it('passes with name that exists', () => {
      expect(validateName(TEST_NAME)).toBe(true);
    });
  });

  describe('aria-hidden', () => {
    it('is true when there is no aria-label', () => {
      createComponent();

      expect(wrapper.attributes('aria-hidden')).toBe('true');
    });

    it('does not exist there is an aria-label', () => {
      createComponent({ 'aria-label': 'Icon label' });

      expect(wrapper.attributes('aria-hidden')).toBeUndefined();
    });
  });
});
