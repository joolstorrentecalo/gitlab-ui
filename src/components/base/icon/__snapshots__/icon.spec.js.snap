// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component when created shows svg class "s8" and path "/path/to/icons.svg#check-circle" 1`] = `
<svg
  aria-hidden="true"
  class="gl-icon s8"
  data-testid="check-circle-icon"
  role="img"
>
  <use
    href="/path/to/icons.svg#check-circle"
  />
</svg>
`;
